File: cli/xbin/doctor.py
"""xbin doctor: check that all required and optional tools are installed."""

from __future__ import annotations

import json
import platform
import shutil
import subprocess
import sys
from pathlib import Path

# ---------------------------------------------------------------------------
# Checks
# ---------------------------------------------------------------------------


def _check(name: str, cmd: list[str] | None = None, hint: str = "") -> tuple[bool, str]:
    """Run a check. Returns (ok, detail)."""
    if cmd is None:
        path = shutil.which(name)
        if path:
            return True, path
        return False, f"{name} not found"
    try:
        r = subprocess.run(cmd, capture_output=True, text=True, timeout=5)
        if r.returncode == 0:
            out = r.stdout.strip().splitlines()[0] if r.stdout.strip() else "ok"
            return True, out
        return False, (
            r.stderr.strip().splitlines()[0]
            if r.stderr.strip()
            else f"exit {r.returncode}"
        )
    except FileNotFoundError:
        return False, f"{cmd[0]} not found"
    except subprocess.TimeoutExpired:
        return False, "timed out"


def _check_musl_target() -> tuple[bool, str]:
    try:
        r = subprocess.run(
            ["rustup", "target", "list", "--installed"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        target = f"{platform.machine()}-unknown-linux-musl"
        if target in r.stdout:
            return True, target
        return False, f"{target} not installed (run: rustup target add {target})"
    except FileNotFoundError:
        return False, "rustup not found"


def _check_python_version() -> tuple[bool, str]:
    v = sys.version_info
    if v >= (3, 10):
        return True, f"{v.major}.{v.minor}.{v.micro}"
    return False, f"{v.major}.{v.minor}.{v.micro} (need >= 3.10)"


def _check_package(name: str) -> tuple[bool, str]:
    try:
        mod = __import__(name)
        ver = getattr(mod, "__version__", "?")
        return True, ver
    except ImportError:
        return False, f"not installed (pip install {name})"


def _check_stub() -> tuple[bool, str]:
    from ._util import find_binary

    try:
        p = find_binary("xbin-stub", "XBIN_STUB", "")
        return True, str(p)
    except FileNotFoundError:
        return False, "not built (run: make stub)"


def _check_crypto() -> tuple[bool, str]:
    from ._util import find_binary

    try:
        p = find_binary("xbin-crypto", "XBIN_CRYPTO", "")
        return True, str(p)
    except FileNotFoundError:
        return False, "not built (run: make stub)"


def _collect_checks() -> list[dict]:
    """Run all checks and return structured results."""
    checks: list[tuple[str, bool, str, bool]] = []

    def add(name: str, ok: bool, detail: str, required: bool = True) -> None:
        checks.append((name, ok, detail, required))

    # --- Required ---
    add("Python", *_check_python_version())
    add(
        "pip",
        *(
            (
                True,
                (
                    subprocess.run(
                        [sys.executable, "-m", "pip", "--version"],
                        capture_output=True,
                        text=True,
                        timeout=5,
                    )
                    .stdout.strip()
                    .split()[1]
                    if subprocess.run(
                        [sys.executable, "-m", "pip", "--version"],
                        capture_output=True,
                        timeout=5,
                    ).returncode
                    == 0
                    else (False, "not available")
                ),
            )
        ),
    )
    add("cargo", *_check("cargo", ["cargo", "--version"]))
    add("rustc", *_check("rustc", ["rustc", "--version"]))
    add("musl target", *_check_musl_target())
    add("C compiler", *_check("cc", ["cc", "--version"]))
    add("zstd", *_check("zstd", ["zstd", "--version"]))

    # --- Optional ---
    add(
        "mksquashfs",
        *_check("mksquashfs", ["mksquashfs", "-version"]),
        required=False,
    )
    add("node", *_check("node", ["node", "--version"]), required=False)
    add("deno", *_check("deno", ["deno", "--version"]), required=False)

    # --- Python packages ---
    add("cryptography", *_check_package("cryptography"), required=False)
    add("ruff", *_check_package("ruff"), required=False)
    add("black", *_check_package("black"), required=False)

    # --- Built binaries ---
    add("xbin-stub", *_check_stub(), required=False)
    add("xbin-crypto", *_check_crypto(), required=False)

    return [
        {"name": name, "ok": ok, "detail": detail, "required": required}
        for name, ok, detail, required in checks
    ]


# ---------------------------------------------------------------------------
# Fixers — one per fixable check
# ---------------------------------------------------------------------------


def _fix_musl_target(verbose: bool) -> tuple[bool, str]:
    target = f"{platform.machine()}-unknown-linux-musl"
    cmd = ["rustup", "target", "add", target]
    if verbose:
        print(f"    $ {' '.join(cmd)}", file=sys.stderr)
    try:
        r = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
        if r.returncode == 0:
            return True, target
        return False, r.stderr.strip().splitlines()[0] if r.stderr else "failed"
    except FileNotFoundError:
        return False, "rustup not found"


def _fix_zstd(verbose: bool) -> tuple[bool, str]:
    cmd = ["sudo", "apt-get", "install", "-y", "zstd"]
    if verbose:
        print(f"    $ {' '.join(cmd)}", file=sys.stderr)
    try:
        subprocess.run(cmd, capture_output=True, text=True, timeout=60)
    except FileNotFoundError:
        return False, "apt-get not found (install zstd manually)"
    return _check("zstd", ["zstd", "--version"])


def _fix_mksquashfs(verbose: bool) -> tuple[bool, str]:
    cmd = ["sudo", "apt-get", "install", "-y", "squashfs-tools"]
    if verbose:
        print(f"    $ {' '.join(cmd)}", file=sys.stderr)
    try:
        subprocess.run(cmd, capture_output=True, text=True, timeout=60)
    except FileNotFoundError:
        return False, "apt-get not found (install squashfs-tools manually)"
    return _check("mksquashfs", ["mksquashfs", "-version"])


def _fix_python_package(name: str, verbose: bool) -> tuple[bool, str]:
    cmd = [sys.executable, "-m", "pip", "install", name]
    if verbose:
        print(f"    $ {' '.join(cmd)}", file=sys.stderr)
    try:
        r = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
        if r.returncode != 0:
            detail = r.stderr.strip().splitlines()[-1] if r.stderr else "pip failed"
            return False, detail
    except FileNotFoundError:
        return False, "pip not found"
    return _check_package(name)


def _fix_stub(verbose: bool) -> tuple[bool, str]:
    repo = Path(__file__).resolve().parents[2]
    makefile = repo / "Makefile"
    if not makefile.exists():
        return False, "Makefile not found (not in xbin repo?)"
    cmd = ["make", "stub"]
    if verbose:
        print(f"    $ {' '.join(cmd)}", file=sys.stderr)
    try:
        r = subprocess.run(
            cmd, capture_output=True, text=True, timeout=300, cwd=str(repo)
        )
        if r.returncode == 0:
            return _check_stub()
        detail = r.stderr.strip().splitlines()[-1] if r.stderr else "build failed"
        return False, detail
    except FileNotFoundError:
        return False, "make not found"


FIXERS: dict[str, tuple[callable, bool]] = {
    "musl target": (_fix_musl_target, False),
    "zstd": (_fix_zstd, False),
    "mksquashfs": (_fix_mksquashfs, False),
    "cryptography": (lambda v: _fix_python_package("cryptography", v), False),
    "ruff": (lambda v: _fix_python_package("ruff", v), False),
    "black": (lambda v: _fix_python_package("black", v), False),
    "xbin-stub": (_fix_stub, False),
    "xbin-crypto": (_fix_stub, False),
}


# ---------------------------------------------------------------------------
# Display helpers
# ---------------------------------------------------------------------------


def _print_checks(results: list[dict], *, verbose: bool) -> bool:
    """Print check results. Returns True if any required check failed."""
    from ._color import green, red, yellow

    required_failed = False
    for r in results:
        marker = " " if r["ok"] else ("X" if r["required"] else "-")
        if r["ok"]:
            status = green(f"[{marker}]")
        elif r["required"]:
            status = red(f"[{marker}]")
        else:
            status = yellow(f"[{marker}]")
        req_label = "" if r["required"] else " (optional)"
        if verbose:
            print(f"  {status:9s} {r['name']:15s} {r['detail']}{req_label}")
        if not r["ok"] and r["required"]:
            required_failed = True
    return required_failed


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


def doctor(
    *,
    verbose: bool = True,
    json_output: bool = False,
    fix: bool = False,
    force: bool = False,
) -> int:
    """Check all prerequisites. Returns 0 if all required checks pass."""
    results = _collect_checks()

    if json_output:
        output = {
            "ok": all(r["ok"] or not r["required"] for r in results),
            "checks": results,
        }
        print(json.dumps(output, indent=2))
        if not fix:
            return 0 if output["ok"] else 1

    required_failed = _print_checks(results, verbose=verbose)

    # No --fix: just report.
    if not fix:
        if required_failed:
            if verbose:
                from ._color import red as _red

                msg = "Some required tools are missing. Install them and re-run."
                print(f"\n{_red(msg)}")
            return 1
        if verbose:
            from ._color import green as _green

            print(f"\n{_green('All required checks passed.')}")
        return 0

    # --fix mode --------------------------------------------------------
    failed = [r for r in results if not r["ok"] and r["required"]]
    if not failed:
        if verbose:
            from ._color import green as _green

            print(f"\n{_green('All required checks passed. Nothing to fix.')}")
        return 0

    fixable = [r for r in failed if r["name"] in FIXERS]
    unfixable = [r for r in failed if r["name"] not in FIXERS]

    if unfixable and verbose:
        from ._color import yellow as _yellow

        names = ", ".join(r["name"] for r in unfixable)
        print(
            f"\n{_yellow(f'Cannot auto-fix: {names}')}",
            file=sys.stderr,
        )
        for r in unfixable:
            print(f"  [-] {r['name']:15s} {r['detail']}", file=sys.stderr)

    if not fixable:
        return 1

    # Confirm if interactive.
    if sys.stdin.isatty() and not force:
        names = ", ".join(r["name"] for r in fixable)
        print(f"\n[xbin] will attempt to fix: {names}", file=sys.stderr)
        try:
            answer = input("Proceed? [y/N] ")
        except (EOFError, KeyboardInterrupt):
            print("", file=sys.stderr)
            return 1
        if answer.lower() not in ("y", "yes"):
            return 1

    from ._color import green as _green
    from ._color import red as _red

    if verbose:
        n = len(fixable)
        print(f"\n[xbin] fixing {n} {'issue' if n == 1 else 'issues'}...\n")

    fixed = 0
    failed_count = 0
    for r in fixable:
        fixer = FIXERS[r["name"]][0]
        if verbose:
            print(f"  [>] {r['name']:15s} fixing...", file=sys.stderr, end="")
        ok, detail = fixer(verbose)
        if ok:
            if verbose:
                print(
                    f"\r  [{_green('OK')}] {r['name']:15s} {detail}",
                    file=sys.stderr,
                )
            r["ok"] = True
            fixed += 1
        else:
            if verbose:
                print(
                    f"\r  [{_red('X')}] {r['name']:15s} {detail}",
                    file=sys.stderr,
                )
            failed_count += 1

    if verbose:
        print()
        if failed_count == 0:
            print(f"{_green('All issues fixed.')}")
        else:
            print(f"{_green(f'{fixed} fixed')}  " f"{_red(f'{failed_count} failed')}")

    if json_output:
        output = {"fixed": fixed, "failed": failed_count, "checks": results}
        print(json.dumps(output, indent=2))

    return 0 if failed_count == 0 else 1

File: xbin-cli/src/commands/doctor.rs
use anyhow::Result;
use clap::Args;
use std::process::Command;

#[derive(Args)]
pub struct DoctorArgs {
    /// Output as JSON
    #[arg(long)]
    pub json: bool,

    /// Quiet output
    #[arg(short, long)]
    pub quiet: bool,

    /// Exit with error if any check fails
    #[arg(long)]
    pub strict: bool,

    /// Attempt to auto-install missing prerequisites
    #[arg(long)]
    pub r#fix: bool,

    /// Skip confirmation prompt for `--fix`
    #[arg(short, long)]
    pub force: bool,
}

struct Check {
    name: String,
    ok: bool,
    detail: String,
    optional: bool,
}

pub fn run(args: DoctorArgs) -> Result<()> {
    let mut checks = vec![
        check_command("python3", &["--version"], false),
        check_command("pip", &["--version"], false),
        check_command("cargo", &["--version"], false),
        check_command("rustc", &["--version"], false),
        check_musl_target(false),
        check_command("cc", &["--version"], false),
        check_command("zstd", &["--version"], false),
        check_xbin_stub(false),
        // Optional checks
        check_command("node", &["--version"], true),
        check_command("deno", &["--version"], true),
        check_command("mksquashfs", &["-version"], true),
        check_python_cryptography(),
    ];

    // Auto-fix missing prerequisites
    if args.r#fix {
        if !args.force {
            eprint!("This will attempt to install missing dependencies. Continue? [y/N] ");
            let mut input = String::new();
            std::io::stdin().read_line(&mut input)?;
            if !input.trim().eq_ignore_ascii_case("y") {
                eprintln!("Aborted");
                return Ok(());
            }
        }

        for check in &mut checks {
            if !check.ok && !check.optional {
                let fix_result = attempt_fix(&check.name);
                match fix_result {
                    Ok(msg) => {
                        check.ok = true;
                        check.detail = msg;
                    }
                    Err(e) => {
                        check.detail = format!("fix failed: {e}");
                    }
                }
            }
        }
    }

    if args.json {
        let items: Vec<_> = checks
            .iter()
            .map(|c| {
                serde_json::json!({
                    "name": c.name,
                    "ok": c.ok,
                    "detail": c.detail,
                    "optional": c.optional,
                })
            })
            .collect();
        println!("{}", serde_json::to_string_pretty(&items)?);
    } else {
        let mut all_ok = true;
        for check in &checks {
            let marker = if check.ok { "✓" } else { "✗" };
            let color = if check.ok { "\x1b[32m" } else { "\x1b[31m" };
            let optional_tag = if check.optional { " (optional)" } else { "" };
            if !args.quiet || !check.ok {
                eprintln!(
                    "  {color}{marker}\x1b[0m {:<20} {}{optional_tag}",
                    check.name, check.detail
                );
            }
            if !check.ok && !check.optional {
                all_ok = false;
            }
        }
        if !args.quiet {
            eprintln!();
            if all_ok {
                eprintln!("All checks passed");
            } else if args.strict {
                anyhow::bail!("Some checks failed — install missing dependencies");
            } else {
                eprintln!("Some checks failed (non-fatal, use --strict to enforce)");
            }
        } else if !all_ok && args.strict {
            anyhow::bail!("Some checks failed");
        }
    }

    Ok(())
}

fn check_command(name: &str, args: &[&str], optional: bool) -> Check {
    match Command::new(name).args(args).output() {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let stderr = String::from_utf8_lossy(&output.stderr);
            let version = if output.status.success() {
                stdout.trim().lines().next().unwrap_or("unknown")
            } else {
                stderr.trim().lines().next().unwrap_or("unknown")
            };
            Check {
                name: name.to_string(),
                ok: output.status.success(),
                detail: version.to_string(),
                optional,
            }
        }
        Err(_) => Check {
            name: name.to_string(),
            ok: false,
            detail: "not found".to_string(),
            optional,
        },
    }
}

fn check_musl_target(optional: bool) -> Check {
    let output = Command::new("rustup")
        .args(["target", "list", "--installed"])
        .output();
    match output {
        Ok(o) if o.status.success() => {
            let stdout = String::from_utf8_lossy(&o.stdout);
            let has_musl = stdout
                .lines()
                .any(|l| l.contains("x86_64-unknown-linux-musl"));
            Check {
                name: "musl target".to_string(),
                ok: has_musl,
                detail: if has_musl {
                    "x86_64-unknown-linux-musl installed".to_string()
                } else {
                    "run: rustup target add x86_64-unknown-linux-musl".to_string()
                },
                optional,
            }
        }
        _ => Check {
            name: "musl target".to_string(),
            ok: false,
            detail: "rustup not found".to_string(),
            optional,
        },
    }
}

fn check_xbin_stub(optional: bool) -> Check {
    match which::which("xbin-stub") {
        Ok(path) => Check {
            name: "xbin-stub".to_string(),
            ok: true,
            detail: path.display().to_string(),
            optional,
        },
        Err(_) => Check {
            name: "xbin-stub".to_string(),
            ok: false,
            detail: "not found (optional)".to_string(),
            optional: true, // always optional
        },
    }
}

fn check_python_cryptography() -> Check {
    let output = Command::new("python3")
        .args(["-c", "import cryptography; print(cryptography.__version__)"])
        .output();
    match output {
        Ok(o) if o.status.success() => {
            let ver = String::from_utf8_lossy(&o.stdout);
            Check {
                name: "cryptography".to_string(),
                ok: true,
                detail: ver.trim().to_string(),
                optional: true,
            }
        }
        _ => Check {
            name: "cryptography".to_string(),
            ok: false,
            detail: "pip install cryptography".to_string(),
            optional: true,
        },
    }
}

fn attempt_fix(name: &str) -> Result<String> {
    match name {
        "musl target" => {
            let status = Command::new("rustup")
                .args(["target", "add", "x86_64-unknown-linux-musl"])
                .status()?;
            if status.success() {
                Ok("installed via rustup".into())
            } else {
                anyhow::bail!("rustup target add failed");
            }
        }
        "cryptography" => {
            let status = Command::new("pip")
                .args(["install", "cryptography"])
                .status()?;
            if status.success() {
                Ok("installed via pip".into())
            } else {
                anyhow::bail!("pip install failed");
            }
        }
        _ => anyhow::bail!("no automatic fix available for '{name}'"),
    }
}
